crates/dojo-core/src/world.cairo

```diff
@@ -567,21 +568,46 @@ mod world {
 
     #[abi(embed_v0)]
     impl UpgradeableState of IUpgradeableState<ContractState> {
-        fn upgrade_state(ref self: ContractState, new_state: Span<felt252>, program_output: Span<felt252>) {
+        fn upgrade_state(ref self: ContractState, new_state: Span<StorageUpdate>, program_output: ProgramOutput) {
             let program_hash = 18468145346606952491117791430885172818429658569709932865894865254677766952;
             let mut da_hasher = PedersenImpl::new(0);
             let mut i = 0;
             loop {
                 if i == new_state.len() {
                     break;
                 }
-                da_hasher = da_hasher.update(*new_state.at(i));
+                da_hasher = da_hasher.update(*new_state.at(i).key);
+                da_hasher = da_hasher.update(*new_state.at(i).value);
                 i += 1;
             };
             let da_hash = da_hasher.finalize();
-            assert(da_hash == *program_output.at(5), 'wrong output hash');
+            assert(da_hash == program_output.world_da_hash, 'wrong output hash');
+
+            let mut program_output_hash = poseidon::PoseidonImpl::new()
```

Comment:
This could effectively be done into a trait that is implemented by the `ProgramOutput`. But we may here have some additional gas with the current behavior I guess, due to the loops?

Reply:
fixed